### ISSUE TITLE
Upgrade fails during database update

### ISSUE THREAD

**Issue by kochs-online** [zenphoto/zenphoto#1342]
Running the setup when upgrading from version 1.5.9 to current Master (1.6RC) fails during conversion of UTF-8 tables to utf8mb4 collation.

debug.log:

> USER ERROR: MySQLi Error: ( `ALTER TABLE zp_options CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_520_ci` ) failed. MySQLi returned the error _Specified key was too long; max key length is 1000 bytes in_ `/[...]/zenphoto/zp-core/classes/class-dbmysqli.php` _on line 80_
> `trigger_error` called from `dbMySQLi->query` (`class-dbmysqli.php` [80])
> from `dbBase->convertTableToUtf8mb4` (`class-dbbase.php` [907])
> from `index.php` [2515]

This is caused by the index `unique_option`. Dropping the index allows setup to convert all tables, but the index cannot be re-created manually.

**Comment by acrylian** [comment]
Didn't get that. That is a combined index of three columns set in line 2305 of setup/index.php. In need missed to update that one somehow.  

If you have time to try add something like this in line 2431 in setup/index.php after all the other index changes if that works:

`$sql_statements[] = "ALTER TABLE $tbl_tags DROP INDEX unique_option, ALTER TABLE $tbl_options ADD UNIQUE `unique_option` (name(191), ownerid, theme(191))"`

**Comment by acrylian** [comment]
(Had a copy&paste type in the first ALTER query, just fixed)

**Comment by kochs-online** [comment]
I already dropped the index to get the site flying, but I tried to execute the SQL statement manually on the database. This error is thrown:

> ERROR 1089 - Incorrect sub part key; the used key part isn't a string, the used length is longer than the key part, or the storage engine doesn't support unique sub keys

**Comment by acrylian** [comment]
But the original query from 1.2.7 to re-create the former one does work?
`"ALTER TABLE $tbl_options ADD UNIQUE `unique_option` (`name`, `ownerid`, `theme`)";`

Or doesn't that work as well?

**Comment by acrylian** [comment]
Well, actually if setup ran before this, it must because setup goes through all of them to not miss any update… I don't know offhand the spec of these special combined type of indexes…

**Comment by acrylian** [comment]
Ah, looking at the other index queries, probably the last part should be like this:

`$sql_statements[] = "ALTER TABLE $tbl_options DROP INDEX unique_option, ALTER TABLE $tbl_options ADD UNIQUE unique_option (name(name(191)), ownerid, theme(theme(191)))`

**Comment by kochs-online** [comment]
> But the original query from 1.2.7 to re-create the former one does work?

Nope, that's throwing the initial error message:

> ERROR 1071: Specified key was too long; max key length is 1000 bytes

To complete the setup I dropped the index and re-ran the second setup step (not sure if I returned to the first step or simply refreshed the page of the second step).

**Comment by kochs-online** [comment]
> `ALTER TABLE $tbl_options ADD UNIQUE unique_option (name(name(191)), ownerid, theme(theme(191)))`

No, that does not work either:

> ERROR 1064 - You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '`name(191)), ownerid, theme(theme(191)))`'

**Comment by acrylian** [comment]
No idea right now, especially since I didn't get any of these errors so far… Will have to try tomorrow, too late today already-

**Comment by acrylian** [comment]
I didn't get the error on my local testsite and I just tested on our live server without getting anything. Which means I am a bit clueless.

**Comment by kochs-online** [comment]
I think we're seeing a limitation of the MyISAM database engine. The MySQL Reference Manual [notes for index prefixes](https://dev.mysql.com/doc/refman/8.0/en/column-indexes.html#column-indexes-prefix):

> Prefixes can be up to 767 bytes long for `InnoDB` tables that use the [`REDUNDANT`](https://dev.mysql.com/doc/refman/8.0/en/glossary.html#glos_redundant_row_format) or [`COMPACT`](https://dev.mysql.com/doc/refman/8.0/en/glossary.html#glos_compact_row_format) row format. The prefix length limit is 3072 bytes for `InnoDB` tables that use the [`DYNAMIC`](https://dev.mysql.com/doc/refman/8.0/en/glossary.html#glos_dynamic_row_format) or [`COMPRESSED`](https://dev.mysql.com/doc/refman/8.0/en/glossary.html#glos_compressed_row_format) row format. For `MyISAM` tables, the prefix length limit is 1000 bytes.

[InnoDB Limits](https://dev.mysql.com/doc/refman/8.0/en/innodb-limits.html) explains:

> The index key prefix length limit is 767 bytes for `InnoDB` tables that use the [`REDUNDANT`](https://dev.mysql.com/doc/refman/8.0/en/glossary.html#glos_redundant_row_format) or [`COMPACT`](https://dev.mysql.com/doc/refman/8.0/en/glossary.html#glos_compact_row_format) row format. For example, you might hit this limit with a [column prefix](https://dev.mysql.com/doc/refman/8.0/en/glossary.html#glos_column_prefix) index of more than 191 characters on a `TEXT` or `VARCHAR` column, assuming a `utf8mb4` character set and the maximum of 4 bytes for each character.
> 
> Attempting to use an index key prefix length that exceeds the limit returns an error.

I'm guessing your test site and you live server are using InnoDB tables in `DYNAMIC` or `COMPRESSED` row format allowing for 3072 bytes of prefix length?

**Comment by acrylian** [comment]
I cannot look quickly into the live db right now but the local one yes. Actually that limitation with the mention 191 lenght I did try to cover with the index dropping/recreation… 

There must be a syntax to limit that problematic combined index the same way somehow. That unique one is of course to so we don't get duplicate options (well, we want some technically for themes as they share the same ones but with their own settings). I guess we can also check that via code (probably we do, would have to look)

**Comment by acrylian** [comment]
(btw, our live server also uses MariaDB but that should not make the difference in general),

**Comment by acrylian** [comment]
I cannot find any example for this but I am probably not looking for hte right term. This didn't work as we know:

     ALTER TABLE $tbl_options ADD UNIQUE unique_option (name(name(191)), ownerid, theme(theme(191)))

Lacking a better idea perhaps the backticks make a difference the original had. Something like that…

     ALTER TABLE $tbl_options ADD UNIQUE unique_option (`name`(name(191)), `ownerid`, `theme`(theme(191)))

Probably not but until we know better perhaps try it nevertheless…

**Comment by kochs-online** [comment]
> (btw, our live server also uses MariaDB but that should not make the difference in general)

Our hoster will switch to MariaDB as well this week (during the same maintenance window they will use to update PHP to version 8.1). But I don't think this will make a difference.



> ```
>  ALTER TABLE $tbl_options ADD UNIQUE unique_option (`name`(name(191)), `ownerid`, `theme`(theme(191)))
> ```

Unfortunately it does not work at all:
> ERROR 1064 - You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'name(191)), \`ownerid\`, \`theme\`(theme(191)))'

**Comment by acrylian** [comment]
Yeah, my mistake (forest for trees… ) as the 2nd part was in any case wrong… But the track was not incorrect in general. My local server accepted this one without error as far as I can see:

           $sql_statements[] = "ALTER TABLE $tbl_options DROP INDEX `unique_option`, ADD UNIQUE `unique_option` (name(191), `ownerid`, theme(191))";

If that char limiting does not throw an error for you it probably cause the too large one for your. So perhaps you have to try values until it works (perhaps 80 for each) . Then we can use that probably.

**Comment by acrylian** [comment]
Using 85 as the limit (of course that's half of 191  ;-) - and "theme" is already smaller with 127) works and via phpmyadmin I see the limit on each column as with the standard indexes:

![Bildschirmfoto 2022-11-14 um 19 30 59](https://user-images.githubusercontent.com/2474309/201738423-6ac7a107-5492-4a09-b526-c0147ac653d1.jpg)

**Comment by kochs-online** [comment]
Doesn't work; I'm getting error 1089 again:

> ERROR 1089 - Incorrect sub part key; the used key part isn't a string, the used length is longer than the key part, or the storage engine doesn't support unique sub keys

Looking closer at the error message and the table structure I noticed column `theme` is only 127 characters wide. That's why MySQL complains about the used length being "_longer than the key part_".

Changing the SQL statement to
```ALTER TABLE zp_options ADD UNIQUE INDEX `unique_option` (name(191), `ownerid`, theme(191))```
brings me back to error 1071 as the total of

- 191 characters of [up to] 4 bytes for name
- 4 bytes of integer for ownerid
- 127 characters of [up to] 4 bytes for theme

is 1276 bytes (exceeding the 1000 byte limit).

To fix this we would need to trim down the index to a total of 249 characters (996 bytes / 4 bytes per character).
Ooooor we could consider converting the `name` column back to `ascii_general_ci` collation because – honestly – who needs [technical] option names to be Unicode anyway?

**Comment by kochs-online** [comment]
Ugh, this just keeps getting better and better…

I just tried to create the unique index using only 122 characters for `name`. The good news is MySQL finally tried to create the index (confirming my previous post), but the bad news is that the index creation failed because the entries would not be unique. :cry:

Running this SQL statement to find duplicates
```
SELECT name, ownerid, theme, COUNT(*) FROM zp_options GROUP BY name, ownerid, theme HAVING COUNT(*) > 1
```
returned a list of 672 entries!

**Comment by acrylian** [comment]
Did you try with 85 instead of 191? 191 - as I mentioned as well - didn't make sense because "theme" is only 127 chars .

           $sql_statements[] = "ALTER TABLE $tbl_options DROP INDEX `unique_option`, ADD UNIQUE `unique_option` (name(85), `ownerid`, theme(85))";

Duplicates in the options table may exists. As said several themes have the same options. Not that many probably but you would have to review them itself. Also there is no general way to cleanup options without the danger of removing unwanted things.

**Comment by kochs-online** [comment]
> Duplicates in the options table may exists. As said several themes have the same options.

The SQL statement above groups by `name`, `ownderid` & `theme` (just like the index). So for all of those 672 [distinct] values there are duplicate entries for the same theme.

I guess this is also not to be expected:
> `SELECT COUNT(*) FROM options`
> | `COUNT(*)` |
> | -- |
> | 127360 |

> `SELECT COUNT(*) FROM options WHERE theme = ''`
> | `COUNT(*)` |
> | -- |
> | 127191 |

**Comment by kochs-online** [comment]
These are the top 10:

> `SELECT name, ownerid, theme, COUNT(*) as count FROM options GROUP BY name, ownerid, theme HAVING COUNT(*) > 1 ORDER BY count DESC`
> `name` | `ownerid` | `theme` | `count`
> -- | -- | -- | --
> `magick_max_height` | 0 |   | 29390
> `magick_max_width` | 0 |   | 29390
> `admin_lastvisit_timeframe` | 0 |   | 25678
> `admin_lastvisit` | 0 |   | 25678
> `graphicslib_selected` | 0 |   | 14697
> `Page-Hitcounter-search` | 0 |   | 540
> `Page-Hitcounter-index` | 0 |   | 73
> `Page-Hitcounter-archive` | 0 |   | 6
> `zp_plugin_security-logger` | 0 |   | 4
> `zp_plugin_rewriteTokens` | 0 |   | 3

**Comment by acrylian** [comment]
I checked with your query on my local and live test site and there are no duplicates. Since you are on a production site with several users that probably also had visitors  these might be from the time the index failed… 

Some of these are set via `setOptionDefault()` which actually should set options that are not yet set.  While the standard `setOption()` has a query addition of `ON DUPLICATE KEY UPDATE` the `setOptionDefault()` seems not to have any check if the option already exists. But the INSERT query would of course correctly fail if that index is set and therefore not add any duplicate. 

So either we have here a long missed incompatibility with MyISAM that is not even related to the utf8mb4 change besides the too long index (you didn't tell yet if the new quere I suggested fixes it) or something else is wrong on your database…

**Comment by kochs-online** [comment]
> Since you are on a production site with several users that probably also had visitors these might be from the time the index failed…

You're right. I just checked the backup I created before starting the update process from version 1.5.4 to 1.5.9. No duplicates.
I'm sorry for having bothered you with that!

> you didn't tell yet if the new quere I suggested fixes it

Which query exactly are you refering to? I tried the one I mentioned earlier:
```
ALTER TABLE options ADD UNIQUE INDEX `unique_option` (name(122), `ownerid`, theme(127))
```
It looks like it might work, but now it fails because of the duplicate entries. I'll have to clean that up first (not looking forward to that!)…

**Comment by acrylian** [comment]
> I'm sorry for having bothered you with that!

No problem, good occasion to check this.

> For index unique_option we need 4 bytes for ownerid which leaves us 763 bytes (or at least 190 utf8mb4 characters) to split between name and theme. Splitting that evenly would allow for 95 characters each:

Right, even if ownerid is not a string column it counts of course here. If 95 works for you we'll take that.

**Comment by acrylian** [comment]
> Which query exactly are you refering to? I tried the one I mentioned earlier:

Just for completeness I meant this one with 85 each: https://github.com/zenphoto/zenphoto/issues/1342#issuecomment-1314270148 But as said if 95 works that's great. I will commit that soon.

**Comment by kochs-online** [comment]
I will test it after the cleanup and report back here.

**Comment by acrylian** [comment]
Okay, thanks for the help so far.

**Comment by kochs-online** [comment]
All the values in options are settings for themes, plugins, core etc., right? So I can safely restore this table from the backup without risking to lose user data?

**Comment by acrylian** [comment]
Yes, that "should" have no other data. User data is in the administrator table plus the admin_to_object table for any specific rights connections.

**Comment by kochs-online** [comment]
Hm, I should have used a less ambiguous term instead of "user data". What I meant was "user generated data" like albums, photos, tags etc.

**Comment by acrylian** [comment]
Yes, those are in other tables, too. Please see also here: https://www.zenphoto.org/news/database-scheme/

**Comment by kochs-online** [comment]
Yes, I looked at the database scheme, but since the `options` table is missing in the scheme I thought I'd rather ask before wiping the table. Better safe than sorry. :wink:

**Comment by acrylian** [comment]
Yes, options is indeed missing and not even mentioned (ineed to add a actual annotated list sometime…) But it has no direct connections like the others in the diagram do. There might be some connections or settings, I cannot tel offhand which exact but it should be none of the main user related things.

I suggest to make a db backup before you restore, just in case…

**Comment by kochs-online** [comment]
> I will test it after the cleanup and report back here.

Tested successfully! :+1:

Now that the table `options` is fixed I took a closer look at other index prefixes that might cause problems. Table `search_cache` has an index named `criteria` that is 250 characters wide currently. This will exceed the maximum length of 767 bytes for `InnoDB` engines using `REDUNDANT` or `COMPACT` row format. Maybe that should be reduced to 191 characters?

**Comment by acrylian** [comment]
Hm, okay, that makes sense. I will limit its index to 191 as well. Somehow I had this already but somehow removed it thinking only varchar type was affected. This encoding stutt is mind bending a bit…

**Comment by acrylian** [comment]
And awesome we got the original issue fixed, even if we stumbled a bit ;-) I have added the index limitation for criteria now.
